python/raydp/torch/estimator.py

```diff
@@ -16,25 +16,98 @@
 #
 
 import inspect
-from typing import Any, Callable, List, NoReturn, Optional, Union
+from typing import Any, Callable, List, NoReturn, Optional, Union, Dict
+from numpy import outer
 
 import torch
-from ray.util.data import MLDataset
-from ray.util.sgd.torch.torch_dataset import TorchMLDataset
-from ray.util.sgd.torch.torch_trainer import TorchTrainer
-from ray.util.sgd.torch.training_operator import TrainingOperator
 from torch.nn.modules.loss import _Loss as TLoss
-from torch.utils.data.dataloader import DataLoader
 
 from raydp.estimator import EstimatorInterface
-from raydp.spark import RayMLDataset
 from raydp.spark.interfaces import SparkEstimatorInterface, DF, OPTIONAL_DF
 
+from ray import train
+from ray.train import Trainer, TrainingCallback, get_dataset_shard
+import ray.data
+from ray.data.dataset import Dataset
+from ray.data.impl.arrow_block import ArrowRow
+
+def train_fun(config):
+    # create model
+    if isinstance(config["model"], torch.nn.Module):
+        model = config["model"]
+    elif callable(config["model"]):
+        model = config["model"](config)
+    else:
+        raise Exception(
+            "Unsupported parameter, we only support torch.nn.Model instance "
+            "or a function(dict -> model)")
+
+    # create optimizer
+    if isinstance(config["optimizer"], torch.optim.Optimizer):
+        # it is the instance of torch.optim.Optimizer subclass instance
+        # rewrite the optimizer
+        optimizer_cls = config["optimizer"].__class__
+        state = config["optimizer"].state_dict()
+        optimizer = optimizer_cls(model.parameters(), lr=0.1)  # lr must pass for SGD
```

Comment:
why is this 0.1 fixed?

Reply:
I have tested this. The `lr` here is a default value, which is used when the optimizer passed in has no `lr`, otherwise it will be updated to the value set by user.

Reply:
How can user set the lr value, is it included in model.parameters()?

Reply:
lr is included in the optimizer passed in by user, and will be updated in Line52 through `state`